src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    {
+        if (!_closing.exchange(true))
+        {
+            // Stop accepting new output and state changes before we disconnect everything.
+            _connection.TerminalOutput(_connectionOutputEventToken);
+            _connectionStateChangedRevoker.revoke();
+
+            // GH#1996 - Close the connection asynchronously on a background
+            // thread.
+            // Since TermControl::Close is only ever triggered by the UI, we
+            // don't really care to wait for the connection to be completely
+            // closed. We can just do it whenever.
+            _asyncCloseConnection();
+
+            {
+                // GH#8734:
+                // We lock the terminal here to make sure it isn't still being
+                // used in the connection thread before we destroy the renderer.
+                // However, we must unlock it again prior to triggering the
+                // teardown, to avoid the render thread being deadlocked. The
+                // renderer may be waiting to acquire the terminal lock, while
+                // we're waiting for the renderer to finish.
+                auto lock = _terminal->LockForWriting();
+            }
+
+            if (auto localRenderEngine{ std::exchange(_renderEngine, nullptr) })
+            {
+                if (auto localRenderer{ std::exchange(_renderer, nullptr) })
+                {
+                    localRenderer->TriggerTeardown();
+                    // renderer is destroyed
+                }
+                // renderEngine is destroyed
+            }
+
+            // we don't destroy _terminal here; it now has the same lifetime as the
+            // control.
+        }
+    }
+
+    IDXGISwapChain1* ControlCore::GetSwapChain() const
+    {
+        return _renderEngine->GetSwapChain().Get();
+    }
+
+    void ControlCore::_rendererWarning(const HRESULT hr)
+    {
+        auto args{ winrt::make_self<RendererWarningArgs>(hr) };
+        _RendererWarningHandlers(*this, *args);
+    }
+
+    void ControlCore::_renderEngineSwapChainChanged()
+    {
+        _SwapChainChangedHandlers(*this, nullptr);
+    }
+
+    void ControlCore::BlinkAttributeTick()
+    {
+        auto lock = _terminal->LockForWriting();
+
+        auto& renderTarget = *_renderer;
+        auto& blinkingState = _terminal->GetBlinkingState();
+        blinkingState.ToggleBlinkingRendition(renderTarget);
+    }
+
+    void ControlCore::BlinkCursor()
+    {
+        if (!_terminal->IsCursorBlinkingAllowed() &&
+            _terminal->IsCursorVisible())
+        {
+            return;
+        }
+        _terminal->SetCursorOn(!_terminal->IsCursorOn());
+    }
+
+    bool ControlCore::CursorOn() const
+    {
+        return _terminal->IsCursorOn();
+    }
+
+    void ControlCore::CursorOn(const bool isCursorOn)
+    {
+        _terminal->SetCursorOn(isCursorOn);
+    }
+
+    void ControlCore::ResumeRendering()
+    {
+        _renderer->ResetErrorStateAndResume();

```

Comment:
hmm can you confirm the renderer crash state still works?

Reply:
You know, I used to be able to consistently repro this scenario by killing dwm. Now I can't get it to repro at all in any of the windows. Maybe I need a build running while killing dwm?

Reply:
future me: this is `dxcap -forcetdr`, and no it doesn't work right

Reply:
![image](https://user-images.githubusercontent.com/18356694/115616179-7de6f380-a2b5-11eb-98e5-ff347c27f674.png)

nevermind it works just fine